mcf-browser/src/back-end/graph.js

```diff
@@ -0,0 +1,318 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * Module contains Node and Assertion classes which together create a local
+ * version of the Data Commons Knowledge Graph.
+ */
+
+import {
+  getExistsInKG,
+  getRemotePropertyLabels,
+  getRemotePropertyValues,
+  getValueFromValueObj,
+} from './utils.js';
+
+/** Class representation of a single Node in the KG. */
+class Node {
+  /**
+   * Universal id of the node. Every Node obj has a localId, which is either a
+   * local reference used in an uplaoded file, or the dcid if the node is never
+   * referred to locally (remote only Node).
+   * @type {string}
+   */
+  localId;
+  /**
+   * Whether triples from the remote Data Commons Knowledge Graph have already
+   * been fetched.
+   * @type {boolean}
+   */
+  alreadyFetched;
+  /**
+   * Whether the node exists in the Data Commons Knowledge Graph.
+   * @type {boolean}
+   */
+  existsInKG;
+  /**
+   * Dcid of the node. Set only if a remote id is referred to in a local file or
+   * pulled from the Data Commons Knowledge Graph.
+   * @type {?string}
+   */
+  dcid;
+  /**
+   * The head of a linked list of Assertion objects representing the outgoing
+   * triples of the Node object.
+   * @type {?Assertion}
+   */
+  assertions;
+  /**
+   * The head of a linked list of Assertion objects representing the incoming
+   * triples of the Node object.
+   * @type {?Assertion}
+   */
+  invAssertions;
+
+  /**
+   * Create a Node based on a given id.
+   * @param {string} id The id of the node to create.
+   */
+  constructor(id) {
+    this.localId = id;
+    this.alreadyFetched = false;
+    this.existsInKG = false;
+    this.dcid = null;
+    this.assertions = null;
+    this.invAssertions = null;
+  }
+
+  /**
+   * Indicates if a given object is an instance of Node class.
+   * @param {Object} obj The object to check.
+   * @return {boolean} True if the object is an instance of Node.
+   */
+  static isNode(obj) { return obj instanceof Node; }
+
+  /**
+   * Returns a node with the given ID, creates a new node if shouldCreate is
+   * true, or null if the node does not exist and a node should not be created.
+   *
+   * @param {string} id The id of the node to find.
+   * @param {boolean} shouldCreate True if a new Node should be created if it
+   *     does not already exist.
+   * @return {Node|null} The found node if it exists or is created.
+   */
+  static getNode(id, shouldCreate) {
```

Comment:
Which getNode caller passes shouldCreate=false? (My search doesn't find a non-test caller)

Reply:
there had been some front end callers that had called it with false, but I recently changed the implementation so that getNode is never actually called with shouldCreate=false. I'll remove the unused arg, thanks for pointing this out!